mapbox/mapbox-unity-sdk
# Styling improvement

**Related issue**

Example: Closes #832. Relates to #832.

**Description of changes**

Styling UX improvements:

VectorLayers in AbstractMap use a new UI for styling which allows for easy selection of Mapbox sample styles and creation/selection of custom user-created styles. 

Details:

- ScriptableStyle is a scriptableObject which wraps a public reference for GeometryMaterialOptions
- MapFeatureStyleOptions is a new class which wraps public references for StyleType enum and a ScriptableStyle
- MapFeatureStyleOptionsDrawer contains functionality for drawing MapFeatureStyleOptions UI and creating a new style.
- VectorSubLayerProperties, VectorLayerVisualizer and VectorLayerPropertiesDrawer have been updated to reference MapFeatureStyleOptions instead of GeometryMaterialOptions
- MapboxDefaultStyles contains a method called GetGeometryMaterialOptions, which accepts and parses a MapFeatureStyleOptions argument and returns a resulting GeometryMaterialOptions object


Known issues:
- styles are currently designed for extruded buildings; should be usable for any vector layer type that results in rendered geometry with varying results
- fantasy style textures are not complete yet; will be completed and updated prior to release.

Testing:

1. In any scene with an abstract map, create a vector layer visualizer for rendering buildings if one does not already exist.
2. Go through the different sample style types (simple, realistic, fantasy) and make sure icon and description text updates.
3. Test each sample style by pressing play in the editor.
4. Select "custom' style; click 'Create New Style', specify a name, and click create. A new scriptable style should get created, saved in the project, and linked to the 'Custom Style'.
5. Test the custom style by pressing play; buildings should render using the colorful atlas template texture shown below.
6. Add a new visualizer to the vector section of AbstractMap, set to render buildings, repeat above steps.

![image](https://user-images.githubusercontent.com/36168942/39327875-406c6594-494e-11e8-9bc7-a077c79bc363.png)


**QA checklists**

- [x] Add relevant code comments. Every API class and method should have `<summary>` description as well as description of parameters.
- [x] **Add tests for new/changed/updated classes and methods!!!**
- [x] Check out conventions in [CONTRIBUTING.md](https://github.com/mapbox/mapbox-unity-sdk/blob/develop/CONTRIBUTING.md).
- [x] Check out conventions in [CODING-STYLE.md](https://github.com/mapbox/mapbox-unity-sdk/blob/develop/CODING-STYLE.md)
- [x] Update the [changelog](https://github.com/mapbox/mapbox-unity-sdk/blob/develop/documentation/docs/05-changelog.md)

**Reviewers**

@atripathi-mb @apavani @jim-martin
